mgr/dashboard: RBD management (frontend)

ricardoasmarques opened the pull request:
This pull request introduces RBD images and snapshots management.

**1- List RBD images and snapshots**
All RBD images are now displayed on a single table, and you can see RBD details and snapshots by selecting a row.

![screenshot from 2018-04-12 16-59-29](https://user-images.githubusercontent.com/14297426/38689494-bd2aaf4a-3e73-11e8-9689-516ab12c3398.png)

**2- RBD image management**
It's now possible to **Create**, **Edit** and **Delete** RBD images.

![screenshot from 2018-04-12 17-00-06](https://user-images.githubusercontent.com/14297426/38689742-4a87952e-3e74-11e8-9ecc-919e6afe92a8.png)

**3- RBD snapshots management**
It's now possible to **Create**, **Edit** and **Delete** RBD snapshots.

![screenshot from 2018-04-12 17-00-27](https://user-images.githubusercontent.com/14297426/38689791-650765f0-3e74-11e8-82f7-7c7655a795eb.png)

**4- Asynchronous tasks**
By using the `Task Manager` infrastructure, requests will be executed asynchronously (if needed), preventing the UI from blocking or receiving timeouts.

![screenshot from 2018-04-12 17-01-47](https://user-images.githubusercontent.com/14297426/38689900-a92d2e0e-3e74-11e8-94ef-adcc4d4a4f08.png)

**5- TODO**
Features that will be added next:
- Protect/Unprotect RBD snapshot (**DONE**)
- Clone RBD snapshot
- Rollback RBD snapshot (**DONE**)
- Copy RBD image

rjfd commented:
@dillaman could you try this PR (plus the backend PR #21360) and tells us what you think about the user interface?

dillaman commented:
@rjfd definitely -- I merged the last PR into one of my builds and was sad when I realized it was only the backend and the frontend was very upset.

dillaman commented:
... since I'm pretty ignorant of the dashboard build environment, am I doing something wrong since I see the following:

> ERROR in src/app/ceph/block/rbd-form/rbd-form.component.html(199,84): : Property 'advancedEnabled' does not exist on type 'RbdFormComponent'.
> src/app/ceph/block/rbd-form/rbd-form.component.html(204,14): : Property 'advancedEnabled' does not exist on type 'RbdFormComponent'.
>

ricardoasmarques commented:
Oops, my fault... It's an erro that only happens on frontend `prod` build, not on `dev` build.

@dillaman this should be fixed now.

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.ts:
```diff
@@ -0,0 +1,454 @@
+              private taskManagerMessageService: TaskManagerMessageService,
+              private notificationService: NotificationService) {
+    this.features = {
+      'deep-flatten': {
```
Nit: it looks like I can check an unchecked ```Deep flatten``` button on the edit pane if it wasn't already enabled.

dillaman replied:
... with the latest update I cannot disable deep-flatten on images where it's already enabled (it should be disable-only).

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,315 @@
+            </cd-helper>
+          </label>
+          <div class="col-sm-9">
+            <select id="dataPool"
```
Nit: might be nice if the data pool drop-down automatically excludes the currently selected image pool.

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,315 @@
+                   class="form-control"
+                   type="text"
+                   formControlName="size"
+                   placeholder="e.g., 10GB"
```
Nit: consistency between giga and gibibyte

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,315 @@
+              </div>
+              <span i18n
+                    class="help-block"
+                    *ngIf="(formDir.submitted || featuresFormGroups.dirty) && featuresFormGroups.hasError('noFeatureSelected')">At least one feature has to be selected</span>
```
Nit: not required to select at least one feature (0 is a valid feature bitmask).

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,315 @@
+          <h2 i18n
+              class="page-header">Advanced</h2>
+
+          <!-- Object Size -->
```
Nit: would it make more sense to use a drop-down for object size and stripe unit since they can only be a few fixed values?

LenzGr replied:
Both object size and stripe unit are drop-down lists now.

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html:
```diff
@@ -93,6 +93,12 @@
+          <li routerLinkActive="active">
+            <a i18n
+               class="dropdown-item"
+               routerLink="/block/rbd">RBDs</a>
```
Nit: why not just ```Images``` instead of ```RBDs```? It's already under the ```Block``` drop-down so the "B" becomes redundant.

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts:
```diff
@@ -0,0 +1,238 @@
+    this.selection = selection;
+  }
+
+  deleteRbdModal() {
```
If this was supposed to pop-up before I deleted an image, it didn't seem to work. Probably a good idea to have a delete confirmation modal when deleting a snapshot as well.

ricardoasmarques replied:
Will be done after https://github.com/ceph/ceph/pull/20899 is merged.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,195 @@
+        const snapshotToUpdate = resultSnapshots.find((snapshot) => {
+          return snapshot.name === executingTask.metadata['snapshot_name'];
+        });
+        console.log('FOUND');
```
Is this need?

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/rbd.service.ts:
```diff
@@ -0,0 +1,81 @@
+import { HttpClient } from '@angular/common/http';
```
This file should be in the API module.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/services.module.ts:
```diff
@@ -24,6 +26,8 @@ import { TaskManagerService } from './task-manager.service';
     TcmuIscsiService,
     ConfigurationService,
     RbdMirroringService,
+    PoolService,
```
This Service is already in the API module.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/shared.module.ts:
```diff
@@ -2,11 +2,17 @@ import { CommonModule } from '@angular/common';
 import { NgModule } from '@angular/core';
 
 import { ApiModule } from './api/api.module';
+import { AuthService } from './api/auth.service';
```
Please remove the changes in this file. Each service should be provided in the respective `api` or `services` module.

ricardoasmarques commented:
@dillaman Thanks for your feedback. I've address all your comments except the deletion confirmation modal that is waiting for an open PR.

@tspmelo I've addressed all you comments, thanks.

FYI, I've also added the suport for **"Protect/Unprotect" snapshots**.

![screenshot from 2018-04-12 22-18-35](https://user-images.githubusercontent.com/14297426/38735957-5657b812-3f22-11e8-94cc-d68dd5d4dc80.png)

This PR has conflicts with the open backend PR, so I've created a branch with all conflicts resolved, that you can use to test - https://github.com/ricardoasmarques/ceph/tree/wip-list-all-rbds-tmp

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,303 @@
+          </div>
+        </div>
+
+        <!-- Use a dedicated pool -->
```
You should not show this option if no data pool can be selected.

ricardoasmarques replied:
From the usability point of view, I'm in favor of not hiding this field.

If no data pool can be selected user will see a "-- No data pool available --" on the dropdown.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,303 @@
+              <option *ngIf="pools !== null && pools.length === 0"
+                      [ngValue]="null">-- No pools available --
+              </option>
+              <option *ngIf="pools !== null && pools.length > 0"
```
Please select the first pool if only one exists.

ricardoasmarques replied:
Done

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,303 @@
+                   class="form-control"
+                   type="text"
+                   formControlName="size"
+                   placeholder="e.g., 10GiB"
```
If I type `100` it get's auto completed to `100MiB` but the placeholder races the expectation that it should be `GiB` instead.

ricardoasmarques replied:
Done

ricardoasmarques commented:
@Devp00l I recommend you to use this testing branch that contains both frontend and backend PRs:

https://github.com/ricardoasmarques/ceph/tree/wip-list-all-rbds-tmp

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+    'rbd/create': new TaskManagerMessage(
+      (metadata) => `Create RBD '${metadata.pool_name}/${metadata.image_name}'`,
+      (metadata) => `RBD '${metadata.pool_name}/${metadata.image_name}'
+                     have been created successfully`,
```
```...has been...```

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+    'rbd/edit': new TaskManagerMessage(
+      (metadata) => `Update RBD '${metadata.pool_name}/${metadata.image_name}'`,
+      (metadata) => `RBD '${metadata.pool_name}/${metadata.image_name}'
+                     have been updated successfully`,
```
```...has been...```

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+    'rbd/delete': new TaskManagerMessage(
+      (metadata) => `Delete RBD '${metadata.pool_name}/${metadata.image_name}'`,
+      (metadata) => `RBD '${metadata.pool_name}/${metadata.image_name}'
+                     have been deleted successfully`,
```
```...has been...```

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}'`,
+      (metadata) => `Snapshot ` +
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}' ` +
+                    `have been created successfully`,
```
```...has been...```

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}'`,
+      (metadata) => `Snapshot ` +
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}' ` +
+                    `have been updated successfully`,
```
```...has been...```

dillaman commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -20,6 +20,74 @@ class TaskManagerMessage {
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}'`,
+      (metadata) => `Snapshot ` +
+                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}' ` +
+                    `have been deleted successfully`,
```
```...has been...```

Devp00l commented:
@ricardoasmarques It's was working after recompile :)

ricardoasmarques commented:
@dillaman I've address all your comments, including the delete confirmation modal.

ricardoasmarques commented:
The backend part of this PR is now merged https://github.com/ceph/ceph/pull/21360.

FYI, I've now added a new commit with the support for "RBD snapshot rollback".

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+        novalidate>
+    <div class="panel panel-default">
+      <div class="panel-heading">
+        <h3 class="panel-title">{{ editing ? 'Edit' : 'Add' }} RBD</h3>
```
The text is not translated, maybe you could implement it like i did it [for RGW](https://github.com/ceph/ceph/pull/21351/files#diff-cdf6482138ddac78221c4010b241b83cR27).

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+                 for="dataPool">
+            Data pool
+            <span class="required"></span>
+            <cd-helper html="Dedicated pool that stores the object-data of the RBD.">
```
You missed to translate it ``<cd-helper i18n-html html="Dedicated ...">``

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+        <!-- Size -->
+        <div class="form-group"
+             [ngClass]="{'has-error': (formDir.submitted || rbdForm.controls.size.dirty) && rbdForm.controls.size.invalid}">
+          <label translate
```
s/translate/i18n/

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+                   class="form-control"
+                   type="text"
+                   formControlName="size"
+                   placeholder="e.g., 10GiB"
```
Translation is missing: ``ì18n-placeholder``

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+        <div class="form-group"
+             [ngClass]="{'has-error': (formDir.submitted || featuresFormGroups.dirty) && featuresFormGroups.invalid}"
+             formGroupName="features">
+          <label class="col-sm-3 control-label"
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+        <!-- Advanced -->
+        <div class="row">
+          <div class="col-sm-12">
+            <a class="pull-right margin-right-md" (click)="advancedEnabled = true" *ngIf="!advancedEnabled">
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+               [ngClass]="{'has-error': (formDir.submitted || rbdForm.controls.stripingUnit.dirty) && rbdForm.controls.stripingUnit.invalid}">
+            <label class="control-label col-sm-3"
+                   for="stripingUnit">
+              Stripe unit<span class="required"
```
Missing translation. This time it's a little bit more difficult to do so.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+                    name="stripingUnit"
+                    class="form-control"
+                    formControlName="stripingUnit">
+                <option [ngValue]="null">-- Select stripe unit --</option>
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+               [ngClass]="{'has-error': (formDir.submitted || rbdForm.controls.stripingCount.dirty) && rbdForm.controls.stripingCount.invalid}">
+            <label class="control-label col-sm-3"
+                   for="stripingCount">
+              Stripe count<span class="required"
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form.component.html:
```diff
@@ -0,0 +1,304 @@
+        <div class="button-group text-right">
+          <cd-submit-button [form]="rbdForm"
+                            (submitAction)="submit()">
+            {{ editing ? 'Update' : 'Create' }}  RBD
```
Missing translation, check [here](https://github.com/ceph/ceph/pull/21351/files#diff-cdf6482138ddac78221c4010b241b83cR27).

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:
```diff
@@ -0,0 +1,79 @@
+              class="btn btn-sm btn-primary"
+              *ngIf="!selection.hasSingleSelection"
+              routerLink="/rbd/add">
+        <i class="fa fa-fw fa-plus"></i>Add
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:
```diff
@@ -0,0 +1,79 @@
+              *ngIf="selection.hasSingleSelection"
+              [ngClass]="{'disabled': selection.first().executing}"
+              routerLink="/rbd/edit/{{ selection.first()?.pool_name }}/{{ selection.first()?.name }}">
+        <i class="fa fa-fw fa-pencil"></i>Edit
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:
```diff
@@ -0,0 +1,79 @@
+        <span class="caret"></span>
+        <span class="sr-only"></span>
+      </button>
+      <ul *dropdownMenu class="dropdown-menu" role="menu">
```
Missing translation. Check all lines below.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:
```diff
@@ -0,0 +1,79 @@
+</ng-template>
+
+<ng-template #usageNotAvailableTooltipTpl>
+  <div [innerHtml]="'Only available for RBD images with <strong>fast-diff</strong> enabled'"></div>
```
Missing translation

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.html:
```diff
@@ -0,0 +1,64 @@
+<cd-table [data]="data"
```
Check for missing translations

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,279 @@
+    this.data = this._merge(this.snapshots, this.executingTasks);
+  }
+
+  _merge(snapshots: RbdSnapshotModel[], executingTasks: ExecutingTask[] = []) {
```
If this function should be private then consider using ``private merge(...)`` here.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rollback-confirmation-modal/rollback-confimation-modal.component.html:
```diff
@@ -0,0 +1,25 @@
+<div class="modal-header">
```
Check for missing translations

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/rbd.service.ts:
```diff
@@ -0,0 +1,110 @@
+@Injectable()
+export class RbdService {
+
+  constructor(private http: HttpClient) {
```
Consider returning Observables here like [this service](https://github.com/ceph/ceph/pull/21351/files#diff-0a729d63f185cb3c722b720d9738bca2R10).

ricardoasmarques replied:
This service is now returning Observables.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/delete-confirmation-modal/delete-confirmation-modal.component.html:
```diff
@@ -0,0 +1,25 @@
+<div class="modal-header">
```
Check for missing translations.

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/shared/empty.pipe.ts:
```diff
@@ -0,0 +1,14 @@
+export class EmptyPipe implements PipeTransform {
+
+  transform(value: any, args?: any): any {
+    return _.isUndefined(value) || value === null ? '-' : value;
```
You can use ``_.isNull(value)`` here.

rjfd commented:
jenkins retest this please

ricardoasmarques commented:
@votdev I've adressed all your comments. Thanks for your review.

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-form/rbd-form-response.model.ts:
```diff
@@ -0,0 +1,10 @@
+export class RbdFormResponseModel {
```
They differ on one field:

    <   features_name: string[];
    ---
    >   features: Array<string> = [];

Devp00l replied:
How about extending both models by a base type?

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,279 @@
+  private merge(snapshots: RbdSnapshotModel[], executingTasks: ExecutingTask[] = []) {
+    const resultSnapshots = _.clone(snapshots);
+    executingTasks.forEach((executingTask) => {
+      if (executingTask.name === 'rbd/snap/create') {
```
Done

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,275 @@
+      const snapshotExecuting = resultSnapshots.find((snapshot) => {
+        return snapshot.name === executingTask.metadata['snapshot_name'];
+      });
+      if (executingTask.name === 'rbd/snap/create') {
```
Could you clean those clauses up? So they profit from searching first?

In short something like this:
```
const name = executingTask.name;
if (snapshotExecuting){
 if(name...)
else if(...)
...
}else if (name.endsWith('create')){
 ...
}
```

rjfd replied:
I think there must be a balance between code duplication and code readability, if you try to factorize everything in order to avoid any code duplication you might end up with code that is completely unreadable. In this case, I'm of the opinion that the string comparison should be done with the full task name.

This will make code maintenance more easier because if in the future we want to change the name of the task we can easily grep by the full task name.

ricardoasmarques replied:
Done

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,279 @@
+    });
+  }
+
+  openEditSnapshotModal() {
```
Done

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.ts:
```diff
@@ -0,0 +1,279 @@
+      });
+  }
+
+  rollback(snapshotName) {
```
Done

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-snapshot-list/rbd-snapshot-list.component.html:
```diff
@@ -0,0 +1,64 @@
+  </div>
+</cd-table>
+
+<!-- TODO: Move to `CellTemplate` so it can be reused -->
```
Done

ricardoasmarques commented:
@Devp00l I've addressed all your comments.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts:
```diff
@@ -0,0 +1,256 @@
+          rbdToDelete.cdExecuting = 'deleting';
+        }
+      } else if (executingTask.name === 'rbd/edit') {
+        const rbdToDelete = resultRBDs.find((rbd) => {
```
(Sorry didn't spot this before)

Could refactor it to something that is similar to [this](https://github.com/ceph/ceph/pull/21385/files#diff-3218aa223ce1c378579e8578ec8e1429R113)?
Also the const `rbdToDelete` is very irritating for the editing part.

ricardoasmarques replied:
Done

votdev commented:
Do you plan to refactor the services according to my comments in another PR?
Finally LGTM.
